### PR TITLE
GPKG: fix json map filewriter

### LINKED ISSUE #52677: Can read but not write GPKG with String(JSON) attribute
### What is the bug or the crash?

QGIS can open a GeoPackage table that has an attribute of GDAL type `String(JSON)`. However, QGIS cannot save this data into a new GeoPackage. The error is:
```
Export to vector file failed.
Error: Unsupported type for field kiinteistotunnuksenSijainti
```
As a workaround saving succeeds by selecting this option

kuva

However, using this option is changing the content of the StringJSON field.

GDAL can create a proper copy with ogr2ogr.

Actually QGIS does not read or at least show the data right, but at least it does not fail on reading. The content of the StringJSON field as ogrinfo shows it is
`kiinteistotunnuksenSijainti (String(JSON)) = { "type": "Point", "coordinates": [ 373563.22899999999, 6679730.9349999996 ] }`
and this is how it appears in the attribute table in QGIS
kuva

### Steps to reproduce the issue

1. Open the attached GeoPackage from "json_property.zip"
2. Try to save to layer with TOC right click - Export - Save Features As...

### Versions

QGIS version | 3.30.0-'s-Hertogenbosch | QGIS code revision | f186b8efe0
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.6.3
PROJ version | 9.2.0
EPSG Registry database version | v10.082 (2023-02-06)
GEOS version | 3.11.2-CAPI-1.17.2
SQLite version | 3.41.1
PDAL version | 2.5.2
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
gml_application_schema_toolbox | 1.3.1
ImportPhotos | 3.0.5
LAStools | 1.4
Qgis2threejs | 2.7.1
QuickWKT | 3.1
rasmover-master | version 0.2
shapetools | 3.4.10
simplewcs2 | 0.2
valuetool | 3.0.15
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
processing | 2.12.99

[json_property.zip](https://github.com/qgis/QGIS/files/11222488/json_property.zip)

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by elpaso**
Fix #52677

### CODE REVIEW

File: src/core/qgsvectorfilewriter.cpp [hunk 1]
```diff
@@ -703,6 +703,18 @@ void QgsVectorFileWriter::init( QString vectorFileName,
             break;
           }
 
+          case QVariant::Map:
+            // handle GPKG conversion to JSON
+            if ( mOgrDriverName == QLatin1String( "GPKG" ) )
```
Comment (rouault) [inline src/core/qgsvectorfilewriter.cpp:708]: you could make it more general by checking the "JSON" string in GDALGetMetadata(GDALGetDriverByName(mOgrDriverName), DMD_CREATIONFIELDDATASUBTYPES, nullptr)

File: src/core/qgsvectorfilewriter.cpp [hunk 2]
```diff
@@ -2923,6 +2935,25 @@ gdal::ogr_feature_unique_ptr QgsVectorFileWriter::createFeature( const QgsFeatur
         //intentional fall-through
         FALLTHROUGH
 
+      case QVariant::Map:
+        // handle GPKG conversion to JSON
+        if ( mOgrDriverName == QLatin1String( "GPKG" ) )
```
Comment (rouault) [inline src/core/qgsvectorfilewriter.cpp:2940]: same as above

File: src/core/qgsvectorfilewriter.cpp (hunk 1 above)
Comment (elpaso) [reply src/core/qgsvectorfilewriter.cpp]: done

File: src/core/qgsvectorfilewriter.cpp (hunk 2 above)
Comment (elpaso) [reply src/core/qgsvectorfilewriter.cpp]: done

**Review by rouault — APPROVED** [review]

**Comment by pigreco** [comment]
please backported to the release-3_28 branch

**Pull request merged**